lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java

```diff
@@ -19,15 +19,363 @@
+  }
+
+  /*
+
+Perform a DFS on the provided AST, and Create an AST of similar structure with changes specific to the
+inner query - outer query dynamics. The resultant AST is supposed to be used in outer query.
+
+Base cases:
+ 1. ast is null => null
+ 2. ast is aggregate_function(table.column) => add aggregate_function(table.column) to inner select expressions,
+          generate alias, return aggregate_function(cube.alias). Memoize the mapping
+          aggregate_function(table.column) => aggregate_function(cube.alias)
+          Assumption is aggregate_function is transitive i.e. f(a,b,c,d) = f(f(a,b), f(c,d)). SUM, MAX, MIN etc
+          are transitive, while AVG, COUNT etc are not. For non-transitive aggregate functions, the re-written
+          query will be incorrect.
+ 3. ast has aggregates - iterate over children and add the non aggregate nodes as is and recursively get outer ast
+ for aggregate.
+ 4. If no aggregates, simply select its alias in outer ast.
+ 5. If given ast is memorized as mentioned in the above cases, return the mapping.
+ */
+  private ASTNode getOuterAST(ASTNode astNode, ASTNode innerSelectAST,
+                              AliasDecider aliasDecider, StorageCandidate sc) throws LensException {
+    if (astNode == null) {
+      return null;
+    }
+    if (isAggregateAST(astNode)) {
+      return processAggregate(astNode, innerSelectAST, aliasDecider);
+    } else if (hasAggregate(astNode)) {
+      ASTNode outerAST = new ASTNode(astNode);
+      for (Node child : astNode.getChildren()) {
+        ASTNode childAST = (ASTNode) child;
+        List<String> msrCols = new ArrayList<>();
+        if (hasAggregate(childAST) && sc.getColumns().containsAll(getMeasureFromAggregeteExpr(childAST, msrCols))) {
+          outerAST.addChild(getOuterAST(childAST, innerSelectAST, aliasDecider, sc));
+        } else if (hasAggregate(childAST) && !sc.getColumns().containsAll
+            (getMeasureFromAggregeteExpr(childAST,msrCols))) {
+          childAST.replaceChildren(1, 1, getDefaultNode(null));
+          outerAST.addChild(getOuterAST(childAST, innerSelectAST, aliasDecider, sc));
+        } else {
+          outerAST.addChild(childAST);
+        }
+        /*if (hasAggregate(childAST)) {
+          outerAST.addChild(getOuterAST(childAST, innerSelectAST, aliasDecider, sc));
+        } else {
+          outerAST.addChild(childAST);
+        }*/
+      }
+      return outerAST;
+    } else {
+      ASTNode innerSelectASTWithoutAlias = MetastoreUtil.copyAST(astNode);
+      ASTNode innerSelectExprAST = new ASTNode(new CommonToken(HiveParser.TOK_SELEXPR, "TOK_SELEXPR"));
+      innerSelectExprAST.addChild(innerSelectASTWithoutAlias);
+      String alias = aliasDecider.decideAlias(astNode);
+      ASTNode aliasNode = new ASTNode(new CommonToken(Identifier, alias));
+      innerSelectExprAST.addChild(aliasNode);
+      innerSelectAST.addChild(innerSelectExprAST);
+      if (astNode.getText().equals("0.0")) {
```

Comment:
can we define "0.0" as a public static final String constant and use everywhere ?